service/src/event/service.rs

```diff
@@ -197,225 +177,88 @@ impl CeramicEventService {
-            insert_now: Vec::with_capacity(items.len()),
-            notify_task_new: Vec::with_capacity(items.len()),
-            background_task_deliverable: HashMap::new(),
-        };
-
-        let mut insert_after_history_check: Vec<(DeliverableMetadata, EventInsertable)> =
-            Vec::with_capacity(items.len());
-
-        for item in items {
-            let (insertable, maybe_prev) = Self::parse_item(item).await?;
-            if let Some(meta) = maybe_prev {
-                insert_after_history_check.push((meta, insertable));
-            } else {
-                let init_cid = insertable.body.cid;
-                result.mark_event_deliverable_now(insertable, init_cid);
-            }
-        }
+        let ordered = OrderEvents::try_new(&self.pool, to_insert).await?;
 
-        trace!(local_events_checking=%insert_after_history_check.len(), "checking local history");
-        result
-            .verify_history_inline(pool, insert_after_history_check)
-            .await?;
-        Ok(result)
-    }
+        let missing_history = ordered
+            .missing_history
+            .iter()
+            .map(|(e, _)| e.order_key.clone())
+            .collect();
 
-    async fn parse_item<'a>(
-        item: &ReconItem<'a, EventId>,
-    ) -> Result<(EventInsertable, Option<DeliverableMetadata>)> {
-        let cid = item.key.cid().ok_or_else(|| {
-            Error::new_invalid_arg(anyhow::anyhow!("EventID is missing a CID: {}", item.key))
-        })?;
-        // we want to end a conversation if any of the events aren't ceramic events and not store them
-        // this includes making sure the key matched the body cid
-        let (insertable_body, maybe_prev) =
-            CeramicEventService::parse_event_carfile(cid, item.value).await?;
-        let insertable = EventInsertable::try_new(item.key.to_owned(), insertable_body)?;
-        Ok((insertable, maybe_prev))
-    }
+        let to_insert_with_metadata = if history_required {
+            ordered.deliverable
+        } else {
+            ordered
+                .deliverable
+                .into_iter()
+                .chain(ordered.missing_history)
+                .collect()
+        };
 
-    fn mark_event_deliverable_later(
-        &mut self,
-        insertable: EventInsertable,
-        meta: DeliverableMetadata,
-    ) {
-        self.background_task_deliverable
-            .insert(insertable.body.cid, meta);
-        self.insert_now.push(insertable);
-    }
+        let to_insert = to_insert_with_metadata
+            .iter()
+            .map(|(e, _)| e.clone())
+            .collect::<Vec<_>>();
 
-    fn mark_event_deliverable_now(&mut self, mut ev: EventInsertable, init_cid: Cid) {
-        ev.set_deliverable(true);
-        self.notify_task_new
-            .push(DeliveredEvent::new(ev.body.cid, init_cid));
-        self.insert_now.push(ev);
-    }
+        let res = CeramicOneEvent::insert_many(&self.pool, &to_insert[..]).await?;
```

Comment:
In reading the code it _seems_ like if we change this method to accept an iterator instead of a slice we do not have to do as much cloning of the data. For example we could keep the to_insert vector as a tuple of the insertable and the metadata and use an iter to hide the metadata. This would avoid having the map the insertable back to its metadata below.

Reply:
yeah, I thought that too but hesitated to add another refactor. I did it in the new branch and it removes some allocations (the clone above as well as collects)